src/lib/opts.rs

```diff
@@ -250,6 +249,65 @@ impl Opts {
         }
         Ok(output_types_to_write)
     }
+
+    /// Builds a new copy of this `Opt` where read structures with variable length sample barcodes
+    /// are replaced with fixed length sample barcodes.  The fixed length of each sample barcode
+    /// is inferred from the FASTQ corresponding to the given read structure by examining the
+    /// length of the first read, then removing any other fixed length segments, with the remaining
+    /// length used as the fixed sample barcode length.
+    ///
+    /// If there is no read structure with a variable length sample barcode, the `Opt` itself is
+    /// returned.
+    pub fn with_fixed_sample_barcodes(self) -> Result<Self> {
+        if self
+            .read_structures
+            .iter()
+            .all(|s| s.sample_barcodes().all(read_structure::ReadSegment::has_length))
+        {
+            Ok(self)
+        } else {
+            // Go through each read structure and try to infer the length of any variable length
+            // sample barcode
+            let mut read_structures: Vec<ReadStructure> = vec![];
+            for (read_structure, fastq) in self.read_structures.iter().zip(self.fastqs.iter()) {
+                if read_structure
+                    .iter()
+                    .all(|s| s.kind != SegmentType::SampleBarcode || s.has_length())
+                {
+                    // No variable length, so just clone the current read structure
+                    read_structures.push(read_structure.clone());
+                } else {
+                    // Get the read length from the FASTQ, subtract all non variable length
+                    // segments (must be a sample barcode if we've gone this far).
+                    let read_length: usize = infer_fastq_sequence_length(fastq.clone())?;
+                    let fixed_length: usize =
+                        read_structure.iter().map(|s| s.length().unwrap_or(0)).sum();
+
+                    let read_segments: Vec<ReadSegment> = read_structure
+                        .iter()
+                        .filter_map(|s| {
+                            if s.has_length() {
+                                Some(*s)
+                            } else if fixed_length == read_length {
+                                None
```

Comment:
Do we really want to allow this?  This seems like it should only happen if there's an error somewhere.  Could check it on line 285 and return an Err instead?

Reply:
I get that `+B` can technically resolve to 0 bases on any given read, but in this particular case wouldn't it be weird to have a read structure that has a `+B` in it and then have that resolve to zero bases?  And then omitting that segment from the read structure will cause super-confusing error messages because it will shift any later `B` segments down by one when comparing to the sample sheet/metadata.

Hence the suggestion to just error out earlier if the fixed-length segments add up to the read length, and we have a disappearing barcode segment.  I'd say you could insert a `0B` here but I see the read-structure crate disallows that (somewhat sensibly).